Added negated property to Relational and BooleanAtom

oscargus opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234". See
https://github.com/blog/1506-closing-issues-via-pull-requests . Please also
write a comment on that issue linking back to this pull request once it is
open. -->
Fixes #15869 

#### Brief description of what is fixed or changed
I have added a property `negated` to `Relational`s that returns the opposite Relational, so `Eq(x, y).negated` gives `Ne(x, y)`. This has only minor benefits over ~, one is that it ignores evaluate = False, so code that relies on negating a relation, primarily in `logic`, see #15869. The purpose of this code is to compare temporary expressions, so not to actually evaluate the expressions, so from that perspective it should be OK, just as the property `canonical` returns a slightly different expression, independent of evaluate. Hence, it should be used with a bit of care.

I also added the property to `BooleanAtom` which makes sense because the result of a `Relational` is a `BooleanAtom`, but doesn't make sense as no other objects in `logic` has that property. From a coding perspective it is only used in a few places (at the moment, only checked quite few cases, could be more).

#### Other comments
Not sure if there should be a release note. I guess it can be convenient for the user to know about it sometimes.

~~I also believe that it is slightly faster than using Not as there should be more processing steps involved with Not. Not a major factor, but no drawback either.~~ Timeit tests show that this is < 10% slower. Not obvious why.

Finally, I split the tests for #8444 as not all of them fails at the moment. Better to have those working is an executed test. Slightly unrelated, but better put it in here as I happened to see it now.

Oh, yeah, I was thinking quite long before deciding on `negated`. One may imagine `inverted` or `complemented` as well. And probably a few more options. I know how to change it...

#### Release Notes

<!-- Write the release notes for this release below. See
https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more information
on how to write release notes. The bot will check your release notes
automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
* core
   * Relationals now have a property negated, which returns the opposite to the relation, similar to ~.
<!-- END RELEASE NOTES -->

oscargus committed:
Added negated property to Relational and BooleanAtom

oscargus commented on sympy/core/relational.py:
```diff
@@ -124,6 +124,28 @@ def reversed(self):
+        """
+        ops = {Eq: Ne, Ge: Lt, Gt: Le, Le: Gt, Lt: Ge, Ne: Eq}
+        a, b = self.args
+        return ops.get(self.func, self.func)(a, b, evaluate=False)
```
The second self.func is a bad idea.

Two alternatives:
1. Live with a NoneType-error if a new Relational is introduced but not handled here.
2. Use a lambda with ~. I have the code, but somehow seems not required as I cannot really see that there are more Relational out there...

smichr replied:
> 1. Live with a NoneType-error if a new Relational is introduced but not handled here.

@asmeurer knows the details better, but maybe this is preferable:

```python
for k in ops:
    if isinstance(self.func, k):
        return ops[k](a, b, evaluate=False)
raise ValueError('unrecognized %s' % self.func)
```

smichr commented:
> deciding on `negated`

it could be `foo.not`

smichr commented on sympy/core/relational.py:
```diff
@@ -124,6 +124,28 @@ def reversed(self):
 
+    @property
+    def negated(self):
+        """Return the negated relationship.
```
you might note here that no evaluation of the result is done

asmeurer commented:
I'm not sure I completely follow the reasoning behind it, but I think it makes sense to have the property. Plus it will be useful if we ever decide to make `Not(Inequality)` be unevaluated.

oscargus commented:
@asmeurer The reasoning is more or less related to what you point out. Some methods in logic relies on comparisons of the type if `(~c1).canonical == c2.canonical`. With `evaluate=False`, `(~c1)` returns a `Not`object and not a `Relational`. The `Not` object does not have a `canonical`property so the code fails with an error.

Related: I was surprised that the new code is actually slower than using ~ (although only in the statistical sense, it varies, and the variance is larger than the difference). However, as we know that the original object is a `Relational`, so the arguments should be "good to go" and that no evaluation should be needed (even if we do not explicitly state it), is there another, more efficient, way to create the new `Relational`?

The corresponding lines in `Not` (in evaluate so the object has been created etc):
 https://github.com/sympy/sympy/blob/1bd5d04ffe2b47b91fd03a868ca31c3c3f9e243e/sympy/logic/boolalg.py#L726-L737

oscargus commented:
> > deciding on `negated`
> 
> it could be `foo.not`

The reason for using a verb is e.g. `reversed`. It sounded more consistent with that and `canonical` (although I realize that it actually should be called `canonicalized` or something...). Also, `is_*`.

It seems like the Python name is actually `invert` so I think my new suggestion is `inverted`. 

(Now, being a non-native English speaker, my feeling may be wrong, but it was primarily for language consistency reasons.)

asmeurer commented:
As far as the name, negated seems good to me. Invert doesn't exactly mean logical negation, and not is a keyword so you can't use it as a method name. 

As for reevaluating the arguments, you should be able to use evaluate=False. There may also be a class method constructor, I'm not sure. There is one on Mul, (_from_args or something like that), but I don't know about other objects.

oscargus committed:
Corrected code and added notes

oscargus committed:
Improved speed slightly for reversed and negated

oscargus commented:
I improved the speed slightly by calling the `__new__` method of `Relational` directly. This decreased the time from 70 to 45 µs (for `Eq`and `Ne`, the other seems faster, now about 15 µs). For `reversed` it reduced from 50 to 45 µs  (again, for `Eq`and `Ne`). I do not fully understand why `negated` was slower than `reversed`earlier, but either way some improvement.

I also added the lambda version of a possibly new `Relational` using ~ as a comment. This will probably help a bit anyway in the unlikely case of a new `Relational`.

Finally, I added some notes trying to outline the difference to `~`/`Not`.

oscarbenjamin commented:
Timings in SymPy are difficult to measure because of caching. Repeating the same measurement gives different results and it's not always clear whether the warm or cold cache time is the one you should worry about it in different situations:

```julia
In [1]: %time s=dsolve(f(x).diff(x)+y**2*x, f(x))                                                                                              
CPU times: user 239 ms, sys: 3.54 ms, total: 243 ms
Wall time: 241 ms

In [2]: %time s=dsolve(f(x).diff(x)+y**2*x, f(x))                                                                                              
CPU times: user 171 ms, sys: 2.41 ms, total: 173 ms
Wall time: 172 ms

In [3]: %time s=dsolve(f(x).diff(x)+y**2*x, f(x))                                                                                              
CPU times: user 170 ms, sys: 2.28 ms, total: 172 ms
Wall time: 171 ms
```

I don't actually know how the cache fits into everything though. Is it possible that a cache is involved in your timings?

When the cache is involved a fair comparison always uses a new process and runs the timings with both cold and warm cache. Of course you don't know whether your cache got warmed up during the import of SymPy which takes a substantial amount of processing:
```console
(venv) enojb@IT035770:~/current/sympy/sympy$ time python -c 'import sympy'

real	0m0.669s
user	0m0.600s
sys	0m0.057s
(venv) enojb@IT035770:~/current/sympy/sympy$ time python -c 'import os'

real	0m0.050s
user	0m0.031s
sys	0m0.012s
```

oscargus commented:
It is also important to test the right thing...

I did `%timeit Eq(x, y).negated`etc. But that also includes the creation of the `Eq` object, which is slower than `Ge` etc.

I cannot really be bothered to run an exhaustive test again, but the actual call to `negated`, assuming an already created object, takes less than 2 µs now. And as the total time, including creating the object has decreased quite a bit more than that, I guess that the relative savings for the change is rather significant.

smichr commented:
`not.negated`
